test/e2e/util/resource_manager.go

```diff
@@ -0,0 +1,316 @@
+package util
+
+import (
+	"context"
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	"github.com/operator-framework/operator-lifecycle-manager/test/e2e/ctx"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	k8scontrollerclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type K8sResourceManager interface {
+	Apply(obj k8scontrollerclient.Object, changeFunc interface{}) error
+	Create(obj k8scontrollerclient.Object, opts ...CreateOption) error
+	Delete(obj k8scontrollerclient.Object, opts ...DeleteOption) error
+	Get(obj k8scontrollerclient.Object, opts ...GetOption) error
+	GetByKey(namespace string, name string, obj k8scontrollerclient.Object, options ...GetOption) error
+	Update(obj k8scontrollerclient.Object) error
+	SetupTestNamespace(name string, options ...TestNamespaceOption) (*corev1.Namespace, error)
+	WaitForConditions(obj k8scontrollerclient.Object, conditions ...ObjectCondition) error
+	List(list k8scontrollerclient.ObjectList, options ...k8scontrollerclient.ListOption) error
+	WaitForEventWithReason(involvedObj k8scontrollerclient.Object, reason string) error
+	Reset(namespacesToDump ...string) error
+}
+
+func NewK8sResourceManager(ctx *ctx.TestContext) K8sResourceManager {
```

Comment:
It's idiomatic in Go -- and I suspect other languages with similar type systems -- to "accept interfaces and return structs", except in certain cases; e.g. abstract factory pattern. This gives a consumer the most flexibility with how it's used.